app/models/transformation_mapping/vm_migration_validator.rb

```diff
@@ -0,0 +1,146 @@
+class TransformationMapping::VmMigrationValidator
+  require 'miq-hash_struct'
+
+  VM_CONFLICT = "conflict".freeze
+  VM_EMPTY_NAME = "empty_name".freeze
+  VM_IN_OTHER_PLAN = "in_other_plan".freeze
+  VM_INVALID = "invalid".freeze
+  VM_MIGRATED = "migrated".freeze
+  VM_NOT_EXIST = "not_exist".freeze
+  VM_VALID = "ok".freeze
+
+  def initialize(mapping, vm_list = nil)
+    @mapping = mapping
+    @vm_list = vm_list
+  end
+
+  def validate
+    @vm_list.present? ? identify_vms : select_vms
+  end
+
+  def select_vms
+    valid_list = []
+
+    vms = Vm.where(:ems_cluster => mapped_clusters).to_a
+    MiqPreloader.preload(vms, %i(lans storages))
+
+    vms.each do |vm|
+      reason = validate_vm(vm, true)
+      valid_list << VmMigrateStruct.new(vm.name, vm, VM_VALID, reason) if reason == VM_VALID
+    end
+
+    {"valid" => valid_list}
+  end
+
+  def identify_vms
+    valid_list = []
+    invalid_list = []
+    conflict_list = []
+
+    vm_names = @vm_list.collect { |row| row['name'] }
+    vm_objects = Vm.where(:name => vm_names, :ems_cluster => mapped_clusters).to_a
+    MiqPreloader.preload(vm_objects, %i(lans storages host ext_management_system))
+
+    @vm_list.each do |row|
+      vm_name = row['name']
+
+      if vm_name.blank?
+        invalid_list << VmMigrateStruct.new('', nil, VM_INVALID, VM_EMPTY_NAME)
+        next
+      end
+
+      vms = vm_objects.select do |vm|
```

Comment:
Wouldn't this be much more efficient to query in the database with a `where` on the `vm_objects` above?